utils/structs/matrix.go

```diff
@@ -0,0 +1,195 @@
+		}
+		n += inc
+
+		for _, v := range m {
+			if inc, err = Vector[T](v).WriteTo(w); err != nil {
+				var t T
+				return n + inc, fmt.Errorf("structs.Vector[%T].WriteTo: %w", t, err)
+			}
+			n += inc
+		}
+
+		return n, w.Flush()
+
+	default:
+		return m.WriteTo(bufio.NewWriter(w))
+	}
+}
+
+// ReadFrom reads on the object from an io.Writer. It implements the
+// io.ReaderFrom interface.
+//
+// If T is a struct, this method requires that T implements io.ReaderFrom.
+//
+// Unless r implements the buffer.Reader interface (see see lattigo/utils/buffer/reader.go),
+// it will be wrapped into a bufio.Reader. Since this requires allocation, it
+// is preferable to pass a buffer.Reader directly:
+//
+//   - When reading multiple values from a io.Reader, it is preferable to first
+//     first wrap io.Reader in a pre-allocated bufio.Reader.
+//   - When reading from a var b []byte, it is preferable to pass a buffer.NewBuffer(b)
+//     as w (see lattigo/utils/buffer/buffer.go).
+func (m *Matrix[T]) ReadFrom(r io.Reader) (n int64, err error) {
+
+	switch r := r.(type) {
+	case buffer.Reader:
+
+		var inc int64
+
+		var size int
+
+		if n, err = buffer.ReadAsUint64[int](r, &size); err != nil {
+			return int64(n), fmt.Errorf("buffer.ReadAsUint64[int]: %w", err)
+		}
+
+		if cap(*m) < size {
+			*m = make([][]T, size)
+		}
+
+		*m = (*m)[:size]
+
+		for i := range *m {
+			if inc, err = (*Vector[T])(&(*m)[i]).ReadFrom(r); err != nil {
+				var t T
+				return n + inc, fmt.Errorf("structs.Vector[%T].ReadFrom: %w", t, err)
+			}
+			n += inc
+		}
+
+		return n, nil
+
+	default:
+		return m.ReadFrom(bufio.NewReader(r))
+	}
+}
+
+// MarshalBinary encodes the object into a binary form on a newly allocated slice of bytes.
+// If T is a struct, this method requires that T implements io.WriterTo.
+func (m Matrix[T]) MarshalBinary() (p []byte, err error) {
+	buf := buffer.NewBufferSize(m.BinarySize())
+	_, err = m.WriteTo(buf)
+	return buf.Bytes(), err
+}
+
+// UnmarshalBinary decodes a slice of bytes generated by
+// MarshalBinary or WriteTo on the object.
+// If T is a struct, this method requires that T implements io.ReaderFrom.
+func (m *Matrix[T]) UnmarshalBinary(p []byte) (err error) {
+	_, err = m.ReadFrom(buffer.NewBuffer(p))
+	return
+}
+
+// Equal performs a deep equal.
+// If T is a struct, this method requires that T implements Equatable.
+func (m Matrix[T]) Equal(other Matrix[T]) bool {
+
+	var t T
+	switch any(t).(type) {
+	case uint, uint64, uint32, uint16, uint8, int, int64, int32, int16, int8, float64, float32:
+
+		if utils.Alias2D[T]([][]T(m), [][]T(other)) {
+			return true
+		}
+
+	default:
+		if _, isEquatable := any(t).(Equatable[T]); !isEquatable {
+			panic(fmt.Errorf("matrix component of type %T does not comply to %T", t, new(Equatable[T])))
+		}
+	}
+
+	isEqual := true
+	for i := range m {
+		isEqual = isEqual && Vector[T](m[i]).Equal(Vector[T](other[i]))
+	}
+
+	return isEqual
```

Comment:
Could return earlier (at first non-equal vector)?
```suggestion
	for i := range m {
		if !Vector[T](m[i]).Equal(Vector[T](other[i])) {
			return false
		}
	}
	return true
```